### PR TITLE
add units module to get the units of features

### PR DESCRIPTION
**Pull request by anilbey**
### Units Module

Added the units API to retrieve the units of the features.

The usage is as follows.
```python
from efel.units import get_unit

assert get_unit("inv_second_ISI") == "Hz"
assert get_unit("ohmic_input_resistance") == "MΩ"
```

### Documentation test

Added a documentation test that parses the eFeatures.rst, extracting the feature name and units, to ensure that the units used in the API remain consistent with any future updates to the documentation.

### Fixes to documentation

Some units in the documentation are corrected with @DrTaDa 's feedback.

### CODE REVIEW

**Review by AurelienJaquier — COMMENTED** [review]
Nice, but the feature file is not complete. There is a list with most of the features in the doc, with units: https://efel.readthedocs.io/en/latest/eFeatures.html

File: efel/units/units.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    "sag_ratio1": "constant",
+    "sag_ratio2": "constant",
+    "ohmic_input_resistance": "mV/nA",
+    "ohmic_input_resistance_vb_ssse": "mV/nA",
+    "voltage_deflection_vb_ssse": "mV",
+    "voltage_after_stim": "mV",
+    "mean_frequency": "Hz",
+    "voltage_deflection": "mV",
+    "voltage_deflection_begin": "mV",
+    "bpo_holding_current": "nA",
```
[the remark this answers is not in the document]
Comment (anilbey) [reply efel/units/units.json:41]: Thanks, I will do that after figuring out the module import issue happening when test is called from tox.

File: efel/units/units.json [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    "AHP_depth": "mV",
+    "AHP_depth_abs": "mV",
+    "AHP_depth_abs_slow": "mV",
+    "AHP_depth_diff": "mV",
+    "AHP_depth_from_peak": "mV",
+    "AHP_depth_last": "mV",
+    "AHP_depth_slow": "mV",
+    "AHP_slow_time": "constant",
+    "AHP_time_from_peak": "ms",
+    "AHP_time_from_peak_last": "mV",
```
Comment (DrTaDa) [inline efel/units/units.json:13]: A time in mV ?

**Comment by anilbey** [comment]
> Nice, but the feature file is not complete. There is a list with most of the features in the doc, with units: https://efel.readthedocs.io/en/latest/eFeatures.html

Alright, the units are updated. There is now a documentation test that parses the rst to fetch the features and units to make sure the API and documentation are consistent. It is part of tox and the CI so whenever a new feature is added it makes sure the API also gets updated.

File: efel/units/units.json (hunk 2 above)
Comment (anilbey) [reply efel/units/units.json:13]: Very good catch, that's why these should be tested. I parsed them programmatically, the error originates here: https://efel.readthedocs.io/en/latest/eFeatures.html#libv5-ahp-time-from-peak-last

File: efel/units/units.json [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    "AP_amplitude_diff": "mV",
+    "AP_amplitude_from_voltagebase": "mV",
+    "AP_begin_time": "ms",
+    "AP_begin_voltage": "mV",
+    "AP_begin_width": "ms",
+    "AP_duration": "ms",
+    "AP_duration_change": "constant",
+    "AP_duration_half_width": "ms",
+    "AP_duration_half_width_change": "constant",
+    "AP_fall_rate": "V",
```
Comment (DrTaDa) [inline efel/units/units.json:29]: Should be V/s no ?

File: efel/units/units.json [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    "AP_fall_rate_change": "constant",
+    "AP_fall_time": "ms",
+    "AP_peak_downstroke": "V",
+    "AP_peak_upstroke": "V",
+    "AP_phaseslope": "1",
+    "AP_phaseslope_AIS": "1",
+    "AP_rise_rate": "V",
```
Comment (DrTaDa) [inline efel/units/units.json:36]: Same ?

File: efel/units/units.json (hunk 2 above)
Comment (anilbey) [reply efel/units/units.json:13]: I'm fixing that one, please double check for more inconsistencies in the features.

File: efel/units/units.json (hunk 4 above)
Comment (anilbey) [reply efel/units/units.json:36]: Good catch, fixing it by updating the parser.

File: efel/units/units.json [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+    "ISI_CV": "constant",
+    "ISI_log_slope": "ms",
+    "ISI_log_slope_skip": "ms",
+    "ISI_semilog_slope": "ms",
+    "ISI_values": "ms",
+    "Spikecount": "constant",
+    "Spikecount_stimint": "constant",
+    "adaptation_index": "constant",
+    "adaptation_index_2": "constant",
+    "all_ISI_values": "Hz",
```
Comment (DrTaDa) [inline efel/units/units.json:56]: Should be ms

File: efel/units/units.json [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    "amp_drop_first_last": "mV",
+    "amp_drop_first_second": "mV",
+    "amp_drop_second_last": "mV",
+    "burst_begin_indices": "constant",
+    "burst_mean_freq": "Hz",
+    "burst_number": "constant",
+    "check_AISInitiation": "constant",
+    "current_base": "mV",
```
Comment (DrTaDa) [inline efel/units/units.json:64]: Should probably be nA not mV for a current

File: efel/units/units.json (hunk 5 above)
Comment (anilbey) [reply efel/units/units.json:56]: Error originates from docs, to be updated

https://efel.readthedocs.io/en/latest/eFeatures.html#libv5-all-isi-values-inv-first-isi-inv-second-isi-inv-third-isi-inv-fourth-isi-inv-fifth-isi-inv-last-isi

File: efel/units/units.json (hunk 6 above)
Comment (anilbey) [reply efel/units/units.json:64]: Error originates from docs, updating...
https://efel.readthedocs.io/en/latest/eFeatures.html#libv5-current-base

File: efel/units/units.json (hunk 2 above)
Comment (anilbey) [reply efel/units/units.json]: Fixed in both docs and API.

File: efel/units/units.json (hunk 4 above)
Comment (anilbey) [reply efel/units/units.json]: Done

File: efel/units/units.json (hunk 5 above)
Comment (anilbey) [reply efel/units/units.json]: Both docs and API are updated

File: docs/source/eFeatures.rst [hunk 7]
```diff
@@ -88,7 +88,7 @@ LibV5 : all_ISI_values, inv_first_ISI, inv_second_ISI, inv_third_ISI, inv_fourth
 1.0 over first/second/third/fourth/fith/last ISI; returns 0 when no ISI
 
 - **Required features**: peak_time (ms)
-- **Units**: Hz
+- **Units**: ms
```
Comment (AurelienJaquier) [inline docs/source/eFeatures.rst:91]: So the problem here is that all_ISI_values is in ms, while the other inv_*_ISI are in Hz. Maybe leave Hz here and create a new entry for all_ISI_values?

File: docs/source/eFeatures.rst [hunk 8]
```diff
@@ -1657,7 +1657,7 @@ The ratio between the voltage deflection and stimulus current
 
 - **Required features**: t, V, stim_start, stim_end, voltage_deflection
 - **Parameters**: stimulus_current
-- **Units**: mV/nA
+- **Units**: MΩ
```
Comment (AurelienJaquier) [inline docs/source/eFeatures.rst:1660]: can greek letters be parsed by sphinx from rst? Have you tried making the docs locally to check that?

File: docs/test_feature_units_in_docs.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+import re
+
+
+from efel import units
+
+
+def extract_feature_and_units(rst_text: str) -> dict:
+    """Parse rst to extract features and units."""
+    # Use regex to match and extract the desired information
+    pattern = r'LibV\d+\s*:\s*([\w_]+)[\s\S]*?- \*\*Units\*\*:\s*([\w/\(\)]+)'
```
Comment (AurelienJaquier) [inline docs/test_feature_units_in_docs.py:15]: I will just trust you with that one ^_^'

File: efel/tests/test_units.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+
+def test_get_unit():
+    """Unit test for the get_unit function."""
+    assert get_unit("time_to_last_spike") == "ms"
+    assert get_unit("inv_second_ISI") == "Hz"
+    assert get_unit("check_AISInitiation") == "constant"
+    assert get_unit("AP1_amp") != "wrong unit"
+    assert get_unit("AP1_amp") == "mV"
+    assert get_unit("ohmic_input_resistance") == "MΩ"
```
Comment (AurelienJaquier) [inline efel/tests/test_units.py:14]: Didn't know that we could just use greek letters in python code. Nice!

File: efel/units/units.json [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    "ISI_values": "ms",
+    "Spikecount": "constant",
+    "Spikecount_stimint": "constant",
+    "adaptation_index": "constant",
+    "adaptation_index_2": "constant",
+    "all_ISI_values": "ms",
+    "amp_drop_first_last": "mV",
+    "amp_drop_first_second": "mV",
+    "amp_drop_second_last": "mV",
+    "burst_begin_indices": "constant",
```
Comment (AurelienJaquier) [inline efel/units/units.json:60]: where did you get value for this? I didn't put the *_indices features in the documentation because it would be confusing for the user due to the interpolation.

Nothing wrong with having them in the units file though.

**Review by AurelienJaquier — COMMENTED** [review]
Could you also add the units/units.json to the files to be updated here?
https://efel.readthedocs.io/en/latest/developers.html#updating-relevant-files
docs/source/developers.rst

File: docs/test_feature_units_in_docs.py (hunk 9 above)
Comment (DrTaDa) [reply docs/test_feature_units_in_docs.py:15]: Holy wow, that's something. 

I ask GPT4 for a translation:
```

ME: What does this regex line do ?
pattern = r'LibV\d+\s*:\s*([\w_]+)[\s\S]*?- \*\*Units\*\*:\s*([\w/\(\)]+)'

GPT4: This regular expression pattern is designed to match and capture specific information from a string. Let's break down the pattern to understand it better:

LibV: This part matches the literal string "LibV".
\d+: This part matches one or more digits (0-9).
\s*: This part matches zero or more whitespace characters (spaces, tabs, newlines, etc.).
:: This part matches the literal colon character.
\s*: This part (again) matches zero or more whitespace characters.
([\w_]+): This part matches one or more word characters (letters, digits, or underscores) and captures them in a group.
[\s\S]*?: This part matches zero or more occurrences of any character, including whitespace and non-whitespace characters, in a non-greedy fashion (i.e., as few characters as possible).
- \*\*Units\*\*:: This part matches the literal string "- Units:".
\s*: This part (again) matches zero or more whitespace characters.
([\w/\(\)]+): This part matches one or more word characters (letters, digits, or underscores), forward slashes, or parentheses, and captures them in a group.
In summary, this regex pattern searches for a string that starts with "LibV" followed by one or more digits, a colon, and one or more word characters or underscores (capturing this part). Then, it matches any number of characters until it finds the string "- Units:" followed by one or more word characters, forward slashes, or parentheses (capturing this part as well).

The pattern is typically used with Python's re module to match and extract information from a larger string. The first captured group would represent a name or identifier, and the second captured group would represent some unit information.

```

File: efel/tests/test_units.py (hunk 10 above)
Comment (DrTaDa) [reply efel/tests/test_units.py:14]: You can put emojis as well

File: docs/source/eFeatures.rst (hunk 8 above)
Comment (anilbey) [reply docs/source/eFeatures.rst:1660]: I tried but couldn't succeed. How do you build the docs?

When I try `make doc`, it proceeds for a while but then it asks me to enter a file name. :thinking: 
```
This is pdfTeX, Version 3.141592653-2.6-1.40.22 (TeX Live 2022/dev/Debian) (preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode
(./efeature-documentation.tex
LaTeX2e <2021-11-15> patch level 1
L3 programming layer <2022-01-21>
(/usr/share/texlive/texmf-dist/tex/latex/base/article.cls
Document Class: article 2021/10/04 v1.4n Standard LaTeX document class
(/usr/share/texlive/texmf-dist/tex/latex/base/size10.clo))
(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsmath.sty
For additional information on amsmath, use the `?' option.
(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amstext.sty
(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsgen.sty))
(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsbsy.sty)
(/usr/share/texlive/texmf-dist/tex/latex/amsmath/amsopn.sty))

! LaTeX Error: File `boxedminipage.sty' not found.

Type X to quit or <RETURN> to proceed,
or enter new name. (Default extension: sty)

Enter file name: 
```

File: efel/units/units.json (hunk 11 above)
Comment (anilbey) [reply efel/units/units.json:60]: I got it from EModelRunner's sscx units.json file. Should I delete it?

File: docs/source/eFeatures.rst (hunk 8 above)
Comment (AurelienJaquier) [reply docs/source/eFeatures.rst:1660]: I usually use tox to do it, but it looks like there is no docs building in the efel tox.

Maybe take inspiration from https://github.com/BlueBrain/EModelRunner/blob/main/tox.ini#L94 to build the docs?

File: efel/units/units.json (hunk 11 above)
Comment (AurelienJaquier) [reply efel/units/units.json:60]: Yes, I guess you can, to be consistent with the docs

File: docs/test_feature_units_in_docs.py (hunk 9 above)
Comment (anilbey) [reply docs/test_feature_units_in_docs.py:15]: Done in the next commit :)

**Comment by anilbey** [comment]
> Could you also add the units/units.json to the files to be updated here? https://efel.readthedocs.io/en/latest/developers.html#updating-relevant-files docs/source/developers.rst

I didn't know about it, it's nice. I just added it there.

**Review by DrTaDa — APPROVED** [review]

File: docs/source/eFeatures.rst (hunk 8 above)
Comment (anilbey) [reply docs/source/eFeatures.rst:1660]: I can try to write something similar to emodelrunner but I am wondering  how is the docs updated in efel in usual? I.e. how did the current docs got built?

File: efel/units/units.json (hunk 11 above)
Comment (anilbey) [reply efel/units/units.json]: Alright it is removed with the latest push.

File: docs/source/eFeatures.rst (hunk 8 above)
Comment (anilbey) [reply docs/source/eFeatures.rst:1660]: The issue was from my system, it was missing this Debian package: texlive-latex-extra
Omega gets rendered fine. The theme of the docs is old btw. let me see if it can be updated

Screenshot_20230428_233459

**Review by AurelienJaquier — APPROVED** [review]

File: docs/source/eFeatures.rst [hunk 12]
```diff
@@ -82,7 +82,19 @@ The time interval between the first too peaks
     doublet_ISI = peak_time[1] - peak_time[0]
 
 
-LibV5 : all_ISI_values, inv_first_ISI, inv_second_ISI, inv_third_ISI, inv_fourth_ISI, inv_fifth_ISI, inv_last_ISI
+LibV5: all_ISI_values
+~~~~~~~~~~~~~~~~~~~~~
+
+The interspike intervals (i.e. time intervals) between adjacent peaks.
```
Comment (DrTaDa) [inline docs/source/eFeatures.rst:88]: This sentence reads a bit weird, maybe:
The interspike intervals, i.e., the time intervals between adjacent peaks.

Comment (anilbey) [reply docs/source/eFeatures.rst]: Thanks, it is now updated.

File: docs/source/eFeatures.rst (hunk 8 above)
Comment (anilbey) [reply docs/source/eFeatures.rst:1672]: I tried the sphinx blue brain theme, it looks like this. The feature names are not in bold however. The rst needs to be updated I think. Let's consider doing that in another PR?
Screenshot_20230501_101259

Comment (AurelienJaquier) [reply docs/source/eFeatures.rst:1672]: sounds good!

**Comment by anilbey** [comment]
The CI did not run docs build. :thinking: . Might it be due to the github actions? @AurelienJaquier  

It runs it locally when I just type `tox` though.
Tox has the following.

```
[tox]
envlist = py3-{style,test},docs
```

Local run:

```
  ❯ tox
  py3-style: OK (57.51=setup[56.91]+cmd[0.02,0.59] seconds)
  py3-test: OK (68.78=setup[33.41]+cmd[0.01,35.37] seconds)
  docs: OK (3.45=setup[0.75]+cmd[2.38,0.32] seconds)
  congratulations :) (129.85 seconds)
```

**Comment by AurelienJaquier** [comment]
That's strange. The command in the github actions is also just `tox`

**Comment by anilbey** [comment]
Alright, now they run and pass.
```

  Build finished. The HTML pages are in build/html.
  docs: commands[1] /home/runner/work/eFEL/eFEL/docs> pytest test_feature_units_in_docs.py
  ============================= test session starts ==============================
  platform linux -- Python 3.7.16, pytest-7.3.1, pluggy-1.0.0
  cachedir: .tox/docs/.pytest_cache
  rootdir: /home/runner/work/eFEL/eFEL
  collected 2 items
  
  test_feature_units_in_docs.py ..                                         [100%]
```

**Comment by anilbey** [comment]
The tests all pass: https://github.com/BlueBrain/eFEL/actions/runs/4849939983.

It appears that old jobs are still marked as "Expected" and "Required" after changes to the workflow configuration file. These lingering checks are preventing the PR from being merged.

**Comment by anilbey** [comment]
Now pip gives HTTP error while installing the dependencies. Luck is not on our side today.

```
ERROR: HTTP error 502 while getting https://files.pythonhosted.org/packages/9e/01/f38e2ff29715251cf25532b9082a1589ab7e4f571ced434f98d0139336dc/pluggy-1.0.0-py2.py3-none-any.whl (from https://pypi.org/simple/pluggy/) (requires-python:>=3.6)
ERROR: Could not install requirement pluggy>=1 from https://files.pythonhosted.org/packages/9e/01/f38e2ff29715251cf25532b9082a1589ab7e4f571ced434f98d0139336dc/pluggy-1.0.0-py2.py3-none-any.whl (from tox) because of HTTP error 502 Server Error: Gateway Error for url: [https://files.pythonhosted.org/packages/9e/01](https://files.pythonhosted.org/packages/9e/01/f38e2ff29715251cf25532b9082a1589ab7e4f571ced434f98d0139336dc/pluggy-1.0.0-py2.py3-none-any.whl)
```

**Comment by anilbey** [comment]
Btw. even if everything runs and the PR gets merged, how is the documentation going to be updated? It's not automated as far as I can see. Does someone have to build and upload it manually?
